energy-dao/src/external_sc_interactions/farm_interactions.rs

```diff
@@ -0,0 +1,252 @@
+        let mut farm_config_mapper = self.farm_config(&farm_address);
+        require!(!farm_config_mapper.is_empty(), ERROR_FARM_DOES_NOT_EXIST);
+        let farm_config = farm_config_mapper.get();
+        require!(
+            farm_config.wrapped_token_id == payment.token_identifier,
+            ERROR_BAD_PAYMENT_TOKENS
+        );
+
+        let current_farm_position = EsdtTokenPayment::new(
+            farm_config.farm_token_id.clone(),
+            farm_config.farm_token_nonce,
+            farm_config.farm_staked_value.clone(),
+        );
+        let claim_rewards_result = self.call_farm_claim(farm_address, current_farm_position);
+        let (new_farm_token, farm_rewards) = claim_rewards_result.into_tuple();
+        self.update_farm_after_claim(
+            &farm_config,
+            &mut farm_config_mapper,
+            new_farm_token.clone(),
+            farm_rewards,
+        );
+
+        let user_rewards = self.compute_user_rewards_payment(&mut farm_config_mapper, &payment);
+
+        farm_config_mapper.update(|config| {
+            config.farm_staked_value -= &payment.amount;
+            config.farm_unstaked_value += &payment.amount;
+            if user_rewards.amount > 0 {
+                config.reward_reserve -= &user_rewards.amount
+            }
+        });
+
+        self.burn_tokens(
+            &payment.token_identifier,
+            payment.token_nonce,
+            &payment.amount,
+        );
+        let current_epoch = self.blockchain().get_block_epoch();
+        let unstake_attributes = UnstakeTokenAttributes {
+            unstake_epoch: current_epoch,
+            token_nonce: new_farm_token.token_nonce,
+        };
+        let unstake_token_payment = self.mint_tokens(
+            farm_config.unstake_token_id.clone(),
+            payment.amount,
+            &unstake_attributes,
+        );
+
+        let caller = self.blockchain().get_caller();
+        self.send_payment_non_zero(&caller, &unstake_token_payment);
+
+        unstake_token_payment
+    }
+
+    #[payable("*")]
+    #[endpoint(unbondFarm)]
+    fn unbond_farm(&self, farm_address: ManagedAddress) -> PaymentsVec<Self::Api> {
+        let payment = self.call_value().single_esdt();
+        let farm_config_mapper = self.farm_config(&farm_address);
+        require!(!farm_config_mapper.is_empty(), ERROR_FARM_DOES_NOT_EXIST);
+        let farm_config = farm_config_mapper.get();
+        require!(
+            farm_config.unstake_token_id == payment.token_identifier,
+            ERROR_BAD_PAYMENT_TOKENS
+        );
+        let unstake_attributes: UnstakeTokenAttributes =
+            self.get_token_attributes(&payment.token_identifier, payment.token_nonce);
+        let current_epoch = self.blockchain().get_block_epoch();
+        let unbond_period = self.unbond_period().get();
+        require!(
+            current_epoch >= unstake_attributes.unstake_epoch + unbond_period,
+            ERROR_UNBOND_TOO_SOON
+        );
```

Comment:
Not a fan of code where I have to spend time thinking what each value means.
```suggestion
        let unbond_period = self.unbond_period().get();
        let unbond_epoch = unstake_attributes.unstake_epoch + unbond_period;
        require!(
            current_epoch >= unbond_epoch,
            ERROR_UNBOND_TOO_SOON
        );
```

Reply:
Updated.